src/cunumeric/scalar_reduction_buffer.h

```diff
@@ -0,0 +1,60 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include "core/cuda/cuda_help.h"
+#include "core/data/buffer.h"
+
+namespace cunumeric {
+
+template <typename REDOP>
+class ScalarReductionBuffer {
+ private:
+  using LHS = typename REDOP::LHS;
+  using RHS = typename REDOP::RHS;
+
+ public:
+  ScalarReductionBuffer(cudaStream_t stream) : buffer_(legate::create_buffer<LHS>(1))
+  {
+    // This will prevent this template from getting instantiated at compile time
+    // if LHS and RHS are different types
+    LHS identity{REDOP::identity};
```

Comment:
This will only emit a narrowing warning if `RHS` (which is the type of `REDOP::identity`) is wider than `LHS`. If `RHS` is not wider then this doesn't do anything. For example, this is compiling fine for me:

```
class TestReduction {
 public:
  typedef float LHS;
  typedef int RHS;
  static const RHS identity;
  template<bool EXCLUSIVE> __CUDA_HD__ static void apply(LHS &lhs, RHS rhs) {}
  template<bool EXCLUSIVE> __CUDA_HD__ static void fold(RHS &rhs1, RHS rhs2) {}
};
const int TestReduction::identity = 0;
...
  int64_t compute_offsets(const AccessorRO<VAL, DIM>& in,
                          const Pitches<DIM - 1>& pitches,
                          const Rect<DIM>& rect,
                          const size_t volume,
                          Buffer<int64_t>& offsets,
                          cudaStream_t stream)
  {
    ScalarReductionBuffer<TestReduction> test(stream);
``` 

Even in the case of narrowing, only a non-fatal warning is emitted, as `legate.mk` is not being built with `-Wfatal-errors`.

Something like `static_assert(std::is_same<LHS, RHS>::value, "...");` might be more robust.

Reply:
alright, I change the code to only use `RHS`, as we only collect RHS values with this buffer (i.e., we never call `apply` on the buffer). So I found it more sensible to use `RHS` as the value type. [14cd060](https://github.com/nv-legate/cunumeric/pull/509/commits/14cd06096a2e9a6974f2955a64b698afe2614bd0)